### PR TITLE
feat: [PL-41387]: GCP OIDC STS, IAM Federal Token Exchange and connector validation

### PR DESCRIPTION
**Pull request by Raghav-Murali**
## Describe your changes

This PR stitches the GCP STS and IAM Access Token exchange. It also adds support to carryout GCP Connector testing for OIDC by ensuring that ID Token generation and federal token exchange can be carried out successfully.

Tested this in local test env.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54564)

### CODE REVIEW

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by tiwaritejas** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by tiwaritejas** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Review by xingchi-jin — APPROVED** [review]

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger ut1, ut4

**Comment by Raghav-Murali** [comment]
trigger ut6

**Comment by Raghav-Murali** [comment]
trigger ut4, ut6

**Comment by Raghav-Murali** [comment]
trigger ut6

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger ut5, ut7

**Comment by Raghav-Murali** [comment]
trigger ut5

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger ut5

**Comment by Raghav-Murali** [comment]
trigger ut5

**Review by jennyjames21 — DISMISSED** [review]

File: 930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcWorkloadAccessTokenResponse.java [hunk 1]
```diff
@@ -21,8 +21,8 @@
 @Data
 @OwnedBy(HarnessTeam.PL)
 public class OidcWorkloadAccessTokenResponse {
-  @JsonProperty(ACCESS_TOKEN) private String accessToken;
-  @JsonProperty(ISSUED_TOKEN_TYPE) private String issuedTokenType;
-  @JsonProperty(TOKEN_TYPE) private String tokenType;
-  @JsonProperty(EXPIRES_IN) private int expiresIn;
+  private String access_token;
+  private String issued_token_type;
```
Comment (gn-harness) [inline 930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcWorkloadAccessTokenResponse.java:25]: why are these not in camel case?

File: 930-oidc-core/src/main/java/io/harness/oidc/gcp/GcpOidcServiceAccountAccessTokenRequest.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import io.harness.annotations.dev.OwnedBy;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.List;
+import lombok.Builder;
+
+@OwnedBy(HarnessTeam.PL)
+@Builder
+public class GcpOidcServiceAccountAccessTokenRequest {
+  @JsonProperty(DELEGATES) private List<String> delegates;
```
Comment (gn-harness) [inline 930-oidc-core/src/main/java/io/harness/oidc/gcp/GcpOidcServiceAccountAccessTokenRequest.java:24]: what does this represent? delegates are the selectors?

File: 930-oidc-core/src/main/java/io/harness/oidc/gcp/GcpOidcTokenUtility.java [hunk 3]
```diff
@@ -88,6 +100,115 @@ public OidcWorkloadAccessTokenResponse exchangeOidcWorkloadAccessToken(
...
+      }
+    } catch (IOException e) {
+      log.error("Exception encountered while exchanging OIDC Service Account Access Token {} ", e);
+    }
+
+    return null;
+  }
+
+  /**
+   * Utility function to generate OIDC ID Token and exchange it for Federal token.
```
Comment (gn-harness) [inline 930-oidc-core/src/main/java/io/harness/oidc/gcp/GcpOidcTokenUtility.java:175]: What is federal token? do you meant federated token? I am not sure if federal token is an alias for the same

File: 930-oidc-core/src/main/java/io/harness/oidc/gcp/GcpOidcTokenUtility.java [hunk 4]
```diff
@@ -88,6 +100,115 @@ public OidcWorkloadAccessTokenResponse exchangeOidcWorkloadAccessToken(
...
+   * If the token exchange happens successfully then do nothing else throw an exception.
+   *
+   * @param workloadPoolId - Workload Identity Pool ID
+   * @param providerId - OIDC Identity Provider ID
+   * @param gcpProjectId - GCP Project ID associated with Workload Identity
+   * @param serviceAccountEmail - Service Account Email with relevant permissions
+   * @param accountId - Harness Account ID
+   */
+  public void validateOidcAccessTokenExchange(
```
Comment (gn-harness) [inline 930-oidc-core/src/main/java/io/harness/oidc/gcp/GcpOidcTokenUtility.java:184]: name sounds strange.. are you exchanging and validating both in the same call?

exchange should be first call and validation should be done afterwards at the call site

File: 930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcWorkloadAccessTokenResponse.java (hunk 1 above)
Comment (Raghav-Murali) [reply 930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcWorkloadAccessTokenResponse.java:25]: Will retain the camel case.

**Comment by Raghav-Murali** [comment]
trigger uts

File: 930-oidc-core/src/main/java/io/harness/oidc/gcp/GcpOidcServiceAccountAccessTokenRequest.java (hunk 2 above)
Comment (Raghav-Murali) [reply 930-oidc-core/src/main/java/io/harness/oidc/gcp/GcpOidcServiceAccountAccessTokenRequest.java:27]: Have added comments to describe this parameter.

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Review by jennyjames21 — APPROVED** [review]

**Pull request merged**